docs/proposals/hpa/federated-hpa.md

```diff
@@ -0,0 +1,135 @@
+---
+title: Federated HPA
+authors:
+- "@YueHonghui" # Authors' github accounts here.
+- "@snowplayfire"
+reviewers:
+- "@robot"
+- TBD
+approvers:
+- "@robot"
+- TBD
+
+creation-date: 2022-02-14
+
+---
+
+# Federated HPA
+## Summary
+HPA is a widely used approach to improve the stablity of apps dealing with burst of request and resource utility. We can run our apps on multiple clusters as easy as a single cluster with Karmada, but Karmada lacks the HPA support now. 
+This proposal introduces a new component FederatedHPA to Karmada and enables Karmada to autoscale workloads across clusters. It also tries to minimize the differences of the user experience between HPA on a single cluster.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.
+-->
+
+### Goals
+1. Bring HPA from single cluster to multiple clusters.
+1. Keep the HPA related CRDs of multi-cluster version the same as the single cluster version.
+1. Tolerate the disaster of member cluster or karmada control plane.
+1. It is better to integrate well with the scenarios such as workloads shifting and cloud burst.
+1. It is better to support different implementations of HPA.
+
+### Non-Goals
+1. Deal the workloads with different subset of services.
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
```

Comment:
i have  a story：

2 member clusters C1 and C2, a deployment D deployed in C1, and the hpa configuration is :

```
target cpu util 30%
min replica 1
max replica 50
```
i hope when the deployment‘s cpu util exceed 30%, the deployment can be scaled up in cluster C1。But if the cluster C1 don't have enough resources for D to scale,  new pods(or deployment) of deployment D can be created in Cluster C2.

Does the design cover this story?

Reply:
@merryzhou Yes, this story is also taken into account in this design. But we may need another component `node autoscaler` to handle this scenario. If cluster C1 don't have enough resources, `node autoscaler` should join more nodes into C1. There is another option that FedHPAController can detect this and redistribute the `min/max`, but I prefer the `node autoscaler` option.

Reply:
Hi， I  have another story：
2 member clusters C1 and C2, a deployment D deployed in C1, and the hpa configuration is :

```
target cpu util 30%
min replica 1
max replica 50
```

When the scaled up conditions are met, I expect there is an hpa strategy to scale the first 30 instances in cluster C1, and the last 20 instances in cluster C2。

Is this story covered？

Reply:
No, this is not covered by this design(FederatedHPA). I think the story is very similar to scenario `cloud burst`, which is that we prefer to use resources in our private cloud than public cloud unless out of capacity. We plan to cover this story in other components not the FederatedHPA. This is exactly what the goal "It is better to integrate well with the scenarios such as workloads shifting and cloud burst" means.